torch_geometric/nn/dense/linear.py

```diff
@@ -258,64 +259,110 @@ def reset_parameters(self):
+            end = time.perf_counter()
+            time_naive = end - start
+
+            # first entry is with segmm, second without
+            # if segmm is faster based on timings, use it
+            self._timing_cache[num_rows] = (time_segmm, time_naive)
+            use_segment_matmul = time_segmm < time_naive
+
+        return use_segment_matmul
+
     def forward(self, x: Tensor, type_vec: Tensor) -> Tensor:
         r"""Forward pass.
 
         Args:
             x (torch.Tensor): The input features.
             type_vec (torch.Tensor): A vector that maps each entry to a type.
         """
-        use_segment_matmul = torch_geometric.backend.use_segment_matmul
-        # If `use_segment_matmul` is not specified, use a simple heuristic to
-        # determine whether `segment_matmul` can speed up computation given the
-        # observed input sizes:
-        if use_segment_matmul is None:
-            if self._use_segment_matmul_heuristic_output is None:
-                segment_count = scatter(torch.ones_like(type_vec), type_vec,
-                                        dim_size=self.num_types, reduce='sum')
-
-                self._use_segment_matmul_heuristic_output = (
-                    torch_geometric.backend.use_segment_matmul_heuristic(
-                        num_segments=self.num_types,
-                        max_segment_size=int(segment_count.max()),
-                        in_channels=self.weight.size(1),
-                        out_channels=self.weight.size(2),
-                    ))
-
-            assert self._use_segment_matmul_heuristic_output is not None
-            use_segment_matmul = self._use_segment_matmul_heuristic_output
-
-        if (use_segment_matmul and torch_geometric.typing.WITH_SEGMM
-                and not is_compiling()):
-            assert self.weight is not None
-
-            perm: Optional[Tensor] = None
-            if not self.is_sorted:
-                if (type_vec[1:] < type_vec[:-1]).any():
-                    type_vec, perm = index_sort(type_vec, self.num_types)
-                    x = x[perm]
-
-            type_vec_ptr = index2ptr(type_vec, self.num_types)
-            out = pyg_lib.ops.segment_matmul(x, type_vec_ptr, self.weight)
-            if self.bias is not None:
-                out += self.bias[type_vec]
-
-            if perm is not None:  # Restore original order (if necessary).
-                out_unsorted = torch.empty_like(out)
-                out_unsorted[perm] = out
-                out = out_unsorted
+        perm: Optional[Tensor] = None
+        if not self.is_sorted:
+            if (type_vec[1:] < type_vec[:-1]).any():
+                type_vec, perm = index_sort(type_vec, self.num_types)
+                x = x[perm]
+
+        type_vec_ptr = index2ptr(type_vec, self.num_types)
+
+        if torch_geometric.backend.use_segment_matmul is None:
+            use_segment_matmul = False
+
+            # TODO check cses of compiling and scripting properly
```

Comment:
~noob question: What does "cses" stand for?~

```suggestion
            # TODO check cases of compiling and scripting properly
```